tianshou/policy/psrl/psrl.py

```diff
@@ -0,0 +1,182 @@
+
+
+class PSRLPolicy(BasePolicy):
+    """Implementation of Posterior Sampling Reinforcement Learning.
+
+    :param PSRLModel model: a model following the rules in
+        :class:`PSRLModel`.
+    :param torch.distributions.Distribution dist_fn: for computing the action.
+    :param float discount_factor: in [0, 1].
+    :param bool reward_normalization: normalize the reward to Normal(0, 1),
+        defaults to ``False``.
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.policy.BasePolicy` for more detailed
+        explanation.
+    """
+
+    def __init__(self,
+                 model: PSRLModel,
+                 dist_fn: torch.distributions.Distribution
+                 = torch.distributions.Categorical,
+                 discount_factor: float = 0.2,
+                 reward_normalization: bool = False,
+                 **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.model = model
+        self.dist_fn = dist_fn
+        assert 0 <= discount_factor <= 1, 'discount factor should in [0, 1]'
+        self._gamma = discount_factor
+        self._rew_norm = reward_normalization
+        self.eps = 0
+
+    def set_eps(self, eps: float) -> None:
+        """Set the eps for epsilon-greedy exploration."""
+        self.eps = eps
+
+    def process_fn(self, batch: Batch, buffer: ReplayBuffer,
+                   indice: np.ndarray) -> Batch:
+        r"""Compute the discounted returns for each frame:
+
+        .. math::
+            G_t = \sum_{i=t}^T \gamma^{i-t}r_i
+
+        , where :math:`T` is the terminal time step, :math:`\gamma` is the
+        discount factor, :math:`\gamma \in [0, 1]`.
+        """
+        # batch.returns = self._vanilla_returns(batch)
+        # batch.returns = self._vectorized_returns(batch)
+        # return batch
+        return self.compute_episodic_return(
+               batch, gamma=self._gamma, gae_lambda=1.)
+
+    def forward(self, batch: Batch,
+                state: Optional[Union[dict, Batch, np.ndarray]] = None,
+                eps: Optional[float] = None,
+                **kwargs) -> Batch:
+        """Compute action over the given batch data.
+
+        :return: A :class:`~tianshou.data.Batch` which has 1 key:
+
+            * ``act`` the action.
+
+        .. seealso::
+
+            Please refer to :meth:`~tianshou.policy.BasePolicy.forward` for
+            more detailed explanation.
+        """
+        act = self.model(batch.obs, state=state, info=batch.info)
+        if eps is None:
+            eps = self.eps
+        if not np.isclose(eps, 0):
+            for i in range(len(act)):
+                if np.random.rand() < eps:
+                    act[i] = np.random.randint(0, self.model.n_action)
+        return Batch(act=act)
+
+    def learn(self, batch: Batch, **kwargs) -> Dict[str, List[float]]:
+        r = batch.returns
+        if self._rew_norm and not np.isclose(r.std(), 0):
+            batch.returns = (r - r.mean()) / r.std()
```

Comment:
this function has moved to self._compute_episodic_return, please refer to the newest master version of PGPolicy.